near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1776 @@
+use std::ops::Bound;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use crate::collections::{append, next_trie_id, Vector};
+use crate::collections::UnorderedMap;
+
+/// TreeMap based on AVL-tree
+///
+/// Runtime complexity (worst case):
+/// - `get`/`contains_key`:     O(1) - UnorderedMap lookup
+/// - `insert`/`remove`:        O(log(N))
+/// - `min`/`max`:              O(log(N))
+/// - `above`/`below`:          O(log(N))
+/// - `range` of K elements:    O(Klog(N))
+///
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct TreeMap<K, V> {
+    root: u64,
+    val: UnorderedMap<K, V>,
+    tree: Vector<Node<K>>,
+}
+
+#[derive(Copy, Clone, BorshSerialize, BorshDeserialize)]
+pub struct Node<K> {
+    id: u64,
+    key: K,             // key stored in a node
+    lft: Option<u64>,   // left link of a node
+    rgt: Option<u64>,   // right link of a node
+    ht: u64,            // height of a subtree at a node
+}
+
+impl<K> Node<K>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize
+{
+    fn of(id: u64, key: K) -> Self {
+        Self {
+            id,
+            key,
+            lft: None,
+            rgt: None,
+            ht: 1,
+        }
+    }
+}
+
+impl<K, V> Default for TreeMap<K, V>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize,
+        V: Copy + BorshSerialize + BorshDeserialize,
+{
+    fn default() -> Self {
+        Self::new(next_trie_id())
+    }
+}
+
+
+impl<K, V> TreeMap<K, V>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize,
+        V: Copy + BorshSerialize + BorshDeserialize,
```

Comment:
I think `V` doesn't need `Copy` at all
And for `K` it makes sense to remove `Copy` and require `Clone` or change some interfaces to `&K`

Reply:
ack

Reply:
PTAL, now `K: Clone + Ord + <ser/de>` and `V: <ser/de>` only.
I was trying to keep `tree` as `Vector<Node<&K>>`, but lost this fight to borrow checker, thus put a few `.clone()`s when moving keys around.